### PR TITLE
Tide: Allow constructing per-org queries

### PR DESCRIPTION
**Pull request by alvaroaleman**
This change allows to get per-org query strings from a tide query. This
is the first step towards adding github apps support for Tide and was
extracted from the full changeset to make reviewing easier.

The old code is left, because we want only want to do this when github
apps auth is used, as it might result in higher token usage.

FTR, the full changeset for tide github apps support that appears to generally be working but needs some prettying and tests is here: https://github.com/kubernetes/test-infra/compare/master...alvaroaleman:tide-github-apps?expand=1

### CODE REVIEW

File: prow/config/tide_test.go
Comment (alvaroaleman) [inline prow/config/tide_test.go:59]: I am btw inclined to change the tests to simply test for a literal search string. It has the drawback that ordering will matter so if someone changes the code in a way that the query is the same but with a different order, those tests will fail incorrectly but:
* It makes the tests much easier to understand (Expected query to be X, was Y)
* It also verifies that there are no additional incorrect parts in the search string, this may introduce bugs that will go unnoticed with the current tests (I had a bug around that initially)

Comment (alvaroaleman) [reply prow/config/tide_test.go:59]: Does anyone have objections to that?

File: prow/config/tide.go
Comment (stevekuznetsov) [inline prow/config/tide.go:266]:
```suggestion
		logrus.WithField("OrgRepoString", orgRepo).Warn("OrgRepo string is invalid, will be ignored")
```

File: prow/config/tide.go
Comment (stevekuznetsov) [inline prow/config/tide.go:272]:
```suggestion
// OrgQueries returns the GitHub search string for the query, sharded
```

(everywhere)

File: prow/config/tide.go
Comment (stevekuznetsov) [inline prow/config/tide.go:216]: Can we GoDoc the output of this func now

File: prow/config/tide.go
Comment (stevekuznetsov) [inline prow/config/tide.go:277]: nit: couldn't this just be

```go
for org, repoIdentifiers := range orgRepoIdentifiers {
    result[org] = queryString + " " + strings.Join(repoIdentifiers, " ")
}
```

Comment (alvaroaleman) [reply prow/config/tide.go:278]: err yeah

File: prow/config/tide_test.go
Comment (alvaroaleman) [reply prow/config/tide_test.go:59]: okay the static string doesn't work because the result string depends on a map order and that is not deterministic. I have added a check for duplicates though

**Comment by alvaroaleman** [comment]
@stevekuznetsov updated, ptal

**Comment by fejta** [comment]
(FYI tide mostly involves clicking the github merge button and is another thing that does not involve gerrit)

File: prow/config/tide.go
Comment (fejta) [inline prow/config/tide.go:267]: Isn't this going to be excessively spammy if something is misconfigured?

File: prow/config/tide_test.go
Comment (fejta) [inline prow/config/tide_test.go:58]: It doesn't look like this covers the case where multiple org scopes are put into the same string? That seems important to cover here.

**Review by fejta — APPROVED** [review]
/hold

File: prow/config/tide_test.go
Comment (alvaroaleman) [reply prow/config/tide_test.go:58]: It does, because the `testQuery` refers to two orgs: `org` and `k`, the latter via the `Repos`/`ExcludedRepos` attributes

File: prow/config/tide.go
Comment (alvaroaleman) [reply prow/config/tide.go]: Yeah, removed if. Ignoring something like that is effectively what already happens today, as github just won't find anything in that case

**Pull request merged**
